pkg/registry/core/service/storage/rest.go

```diff
@@ -337,6 +337,18 @@ func (rs *REST) healthCheckNodePortUpdate(oldService, service *api.Service, node
 func (rs *REST) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc, forceAllowCreate bool, options *metav1.UpdateOptions) (runtime.Object, bool, error) {
 	oldObj, err := rs.services.Get(ctx, name, &metav1.GetOptions{})
 	if err != nil {
+		// Support create on update, if forced to.
+		if forceAllowCreate {
+			obj, err := objInfo.UpdatedObject(ctx, nil)
+			if err != nil {
+				return nil, false, err
+			}
+			createdObj, err := rs.Create(ctx, obj, createValidation, &metav1.CreateOptions{})
```

Comment:
A couple of comments:
1. If this fails on a conflict (because of a race with another create), is it going to be retried by the update/patch handler?
2. You use an empty CreateOptions, are there any options from UpdateOptions that would need to be carried over?

Reply:
> * If this fails on a conflict (because of a race with another create), is it going to be retried by the update/patch handler?

I guess my point is, how is this going to be different from a "regular" storage doing create on update.

Reply:
1. It won't be retried.
2. dry run definitely needs to be carried over. really good catch.

Reply:
> 2. dry run definitely needs to be carried over. really good catch.

I didn't have dry-run in mind, but this would have been a nasty bug!

Reply:
regarding 1, as far as I can tell, the retry loop is already broken by the service custom storage, since it uses the updated object info provided only [once](https://github.com/kubernetes/kubernetes/blob/1a5e48e29d02537828634aac83847aba6aa17c6f/pkg/registry/core/service/storage/rest.go#L356) to create service object to do its special logic on, then it creates a [new updated object info](https://github.com/kubernetes/kubernetes/blob/1a5e48e29d02537828634aac83847aba6aa17c6f/pkg/registry/core/service/storage/rest.go#L425) which always returns that same object and passes that into the retry loop, so if the retry loop fails once, it should be guaranteed to fail on each retry anyway. Basically, I don't think this PR will make the update retry situation any worse for services.